oasis/models/rules.py

```diff
@@ -2,15 +2,151 @@
+from oasis.models.util import sub_id, EPS
+from oasis.libs.features import Features
+from oasis.libs.alert import send_to_slack
+from oasis.libs.rules import parse_rule, OPERATORS
 
 
 class Rules(Model):
-    def __init__(self, job, callback):
+    def __init__(self, job):
+        super(Rules, self).__init__()
         self.job = job
-        self.callback = callback
         self.api = PrometheusAPI(job.data_source)
+        self.timer = Timer(timeparse(self.job.timeout), self.timeout_action)
+        self.config_file = "%s/rules.yml" % self.model_path
+        self.cfg = RulesConfig(self.config_file, job.config)
 
+    def run(self):
+        self.timer.start()
+        for key in self.job.metrics:
+            if key in Metrics:
+                val = Metrics[key]
+                if key not in self.cfg.metrics:
+                    continue
 
+                t = Thread(target=self.run_action,
+                           args=(key, val, self.cfg.metrics[key],
+                                 self.cfg.rules[key]))
+                t.start()
+                self.threads[key] = t
+
+    def run_action(self, metric, val, config, rules):
+        logger.info("[job-id:{id}][metric:{metric}] start ot run"
+                    .format(id=sub_id(self.job.id), metric=metric))
+        self.compute(metric, val, config, rules)
+
+    def compute(self, metric, query_expr, config, rules):
+        # extraction features and match with rules
+        while not self._exit:
+            data_set = self.query_data(query_expr)
+            if len(data_set) > 0:
+                features_value = self.extraction_features(data_set, config)
+                logger.info("[job-id:{id}][metric:{metric}] extraction features {value}, "
+                            "start to match with rule"
+                            .format(id=sub_id(self.job.id),
+                                    metric=metric, value=features_value))
+                self.match_rules(metric, features_value, rules)
+
+            self.event.wait(timeparse(self.cfg.model["predict_interval"]))
+
+        logger.info("[job-id:{id}][metric:{metric}] stop"
+                    .format(id=sub_id(self.job.id), metric=metric))
+
+    def query_data(self, query_expr):
+        now = datetime.datetime.now()
+        query = PrometheusQuery(query_expr,
+                                time.mktime((now - datetime.timedelta(
+                                    seconds=timeparse(self.cfg.model["data_range"])))
+                                            .timetuple()),
+                                time.mktime(now.timetuple()), "15s")
+        return self.api.query(query)
+
+    def extraction_features(self, data_set, config):
+        values = []
+        for data in data_set.values():
+            values.append(float(data))
+
+        features_value = {}
+        for key in config["features"]:
+            if key in Features:
+                features_value[key] = Features[key](values)
```

Comment:
call Features as func?, did we need to eval Features[key]

Reply:
yep, `Features` is a dict of funcs